src/Libraries/Text/Lexer/Tokenizer.idr

```diff
@@ -0,0 +1,148 @@
+match = Match
+
+||| Compose other tokenizer. Language composition should be quoted between
+||| a begin lexer and a end lexer. The begin token can be used to generate
+||| the composition tokenizer and the end lexer.
+export
+compose : (begin : Lexer) ->
+          (mapBegin : String -> a) ->
+          (tagger : String -> tag) ->
+          (middle : Inf (tag -> Tokenizer a)) ->
+          (end : tag -> Lexer) ->
+          (mapEnd : String -> a) ->
+          Tokenizer a
+compose = Compose
+
+||| Stop reason why tokenizer can't make more progress.
+||| @ ComposeNotClosing carries the span of composition begin token in the
+|||                     form of `(startLine, startCol), (endLine, endCol)`.
+public export
+data StopReason = EndInput | NoRuleApply | ComposeNotClosing (Int, Int) (Int, Int)
+
+export
+Show StopReason where
+  show EndInput = "EndInput"
+  show NoRuleApply = "NoRuleApply"
+  show (ComposeNotClosing start end) = "ComposeNotClosing " ++ show start ++ " " ++ show end
+
+export
+Pretty StopReason where
+  pretty EndInput = pretty "EndInput"
+  pretty NoRuleApply = pretty "NoRuleApply"
+  pretty (ComposeNotClosing start end) = "ComposeNotClosing" <++> pretty start <++> pretty end
+
+tokenise : Lexer ->
+           Tokenizer a ->
+           (line, col : Int) -> List (WithBounds a) ->
+           List Char ->
+           (List (WithBounds a), (StopReason, Int, Int, List Char))
+tokenise reject tokenizer line col acc [] = (reverse acc, EndInput, (line, col, []))
+tokenise reject tokenizer line col acc str
+    = case scan reject [] str of
+           Just _ => (reverse acc, (EndInput, line, col, str))
+           Nothing => case getFirstMatch tokenizer str of
+                           Right (toks, line', col', rest) =>
+                               -- assert total because getFirstMatch must consume something
+                               assert_total (tokenise reject tokenizer line' col' (toks ++ acc) rest)
+                           Left reason => (reverse acc, reason, (line, col, str))
+  where
+    countNLs : List Char -> Nat
+    countNLs str = List.length (filter (== '\n') str)
+
+    getCols : List Char -> Int -> Int
+    getCols x c
+        = case span (/= '\n') x of
+               (incol, []) => c + cast (length incol)
+               (incol, _) => cast (length incol)
+
+    getFirstMatch : Tokenizer a -> List Char ->
+                    Either StopReason (List (WithBounds a), Int, Int, List Char)
+    getFirstMatch (Match lex fn) str
+        = let Just (tok, rest) = scan lex [] str
+                | _ => Left NoRuleApply
+              line' = line + cast (countNLs tok)
+              col' = getCols tok col
+              tok' = MkBounded (fn (fastPack (reverse tok))) False line col line' col' in
+              Right ([tok'], line', col', rest)
+    getFirstMatch (Compose begin mapBegin tagger middleFn endFn mapEnd) str
+        = do let Just (beginTok, rest) = scan begin [] str
```

Comment:
Since none of this code is monadic it can be implemented with a single `let` :)

Reply:
Do you like to keep it as it is or to break the rest following `case` as you said above?

Reply:
I think the `case` should be there for sure since the short-circuiting isn't happening right now. It's probably best to clean this up with two `lets`

```
let Just (beginTok, rest) = scan begin [] str
    …
    … in
    case reason of
         (ComposeNotClosing _ _) ⇒ Left reason
         _ ⇒ let …
```

it's not perfect but it removes the repetition, remove the misleading `do-notation` and enable the short-circuiting semantics for `ComposeNotClosing`

Reply:
I personally prefer not to split them even by introducing the repeating `let`, because they are logically in the same layer.

Reply:
Yes that would be ideal but even if you keep the do-notation, you'll have to split the statment in two in order to short-circuit the control flow on `ComposeNotClosing` errors:

```
do let …
   …
   case reason of
        (ComposeNotClosing _ _ _) =>Left reason
        _ ⇒ do let …
```

Reply:
> you'll have to split the statment in two in order to short-circuit the control flow

Why? Seems it's working. What did I miss?

Reply:
It's working but it's not doing what's advertised :) But this is the wrong thread! le me close this and reply on the other